services/management/annotation_test.go

```diff
@@ -0,0 +1,195 @@
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package management
+
+import (
+	"context"
+	"testing"
+
+	"github.com/AlekSi/pointer"
+	"github.com/google/uuid"
+	"github.com/percona/pmm/api/managementpb"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"gopkg.in/reform.v1"
+	"gopkg.in/reform.v1/dialects/postgresql"
+
+	"github.com/percona/pmm-managed/models"
+	"github.com/percona/pmm-managed/utils/logger"
+	"github.com/percona/pmm-managed/utils/testdb"
+	"github.com/percona/pmm-managed/utils/tests"
+)
+
+func TestAnnotations(t *testing.T) {
+	setup := func(t *testing.T) (ctx context.Context, db *reform.DB, teardown func(t *testing.T)) {
+		t.Helper()
+
+		ctx = logger.Set(context.Background(), t.Name())
+		uuid.SetRand(new(tests.IDReader))
+
+		sqlDB := testdb.Open(t, models.SetupFixtures, nil)
+		db = reform.NewDB(sqlDB, postgresql.Dialect, reform.NewPrintfLogger(t.Logf))
+
+		teardown = func(t *testing.T) {
+			uuid.SetRand(nil)
+
+			require.NoError(t, sqlDB.Close())
+		}
+
+		return
+	}
+
+	autorization := []string{"admin:admin"}
+
+	t.Run("Non-existing service", func(t *testing.T) {
+		ctx, db, teardown := setup(t)
+		defer teardown(t)
+		var grafanaClient = new(mockGrafanaClient)
+		s := NewAnnotationService(db, grafanaClient)
+
+		_, err := s.AddAnnotation(ctx, autorization, &managementpb.AddAnnotationRequest{
+			Text:         "Some text",
+			ServiceNames: []string{"no-service"},
+		})
+		require.EqualError(t, errors.New("rpc error: code = NotFound desc = Service with name \"no-service\" not found."), err.Error())
+
+		grafanaClient.AssertExpectations(t)
+	})
+
+	t.Run("Existing service", func(t *testing.T) {
+		ctx, db, teardown := setup(t)
+		defer teardown(t)
+		var grafanaClient = new(mockGrafanaClient)
+		s := NewAnnotationService(db, grafanaClient)
+
+		_, err := models.AddNewService(s.db.Querier, models.MySQLServiceType, &models.AddDBMSServiceParams{
+			ServiceName: "service-test",
+			NodeID:      models.PMMServerNodeID,
+			Address:     pointer.ToString("127.0.0.1"),
+			Port:        pointer.ToUint16(3306),
+		})
+		require.NoError(t, err)
+
+		expectedTags := []string{"service-test"}
+		expectedText := "Some text (Service Name: service-test)"
+		grafanaClient.On("CreateAnnotation", ctx, expectedTags, mock.Anything, expectedText, autorization[0]).Return("", nil)
```

Comment:
Why do we mock it? Essentially, we are testing a mock there.

Reply:
No, we are testing logic in AnnotationService. It's unit tests, not integration.

Reply:
We have separate tests for grafana client to check integration with grafana.

Reply:
I understand your point, but: Does it really have to be a unit test with mocks instead of a real thing? The whole point of introducing a devcontainer was to be able to write tests that use real things, not mocks. We use mocks only we really have too. That philosophy runs in the whole PMM 2.x codebase, so I don't think we suddenly should change that now in a single place.

Reply:
1. What will be difference between API tests and tests in pmm-managed? When should we write API tests and when tests in pmm-managed?
2. We already have places where we use mocks instead of real component(QAN, Prometheus), but we can use real component.
3. We have to export FindAnnotations method in grafana.Client only for tests, which I don't think is a good idea.
According to the description of https://jira.percona.com/browse/PMM-4301, devcontainer was introduced to make it easy to use for everybody.

Reply:
1. We should write API tests if our change affects API, and we should always write tests in pmm-managed repo if that's possible. Tests in pmm-managed repo should not use mocks if possible. In particular, I don't think we should call them "unit tests" as this is a loaded term, and we don't test only "that" package in our tests.
2. Please share examples. To the best of my knowledge, we use mocks in two cases:
* where the usage of a real object would cause import cycles that would require us to shift our code around and make it less organized just for the sake of the tests;
* where we can't trigger some edge case, but we still want to test it (in most cases, we can trigger it and do it);
Both points do not apply there.
3. I semi-agree with that point but don't think it overweight the usage of mocks.

Most importantly, our discussion about changing the usage of mocks should be global, and not a part of a single PR review (but that PR may trigger that discussion).

Reply:
https://github.com/percona/pmm-managed/blob/8e97c4ebaae1388610b5a4af81596bcd467e78f3/services/qan/client_test.go#L128
https://github.com/percona/pmm-managed/blob/8272f31746c5e1d654f895702b97e12ba57ea2e1/services/inventory/agents_test.go#L56
https://github.com/percona/pmm-managed/blob/4b73e3215b39a42c1babc97ecc09b1425fe4a805/services/management/service_test.go#L53

What's the difference between prometheusClient and grafanaClient?

Reply:
Need to check it, but if we need real test, then I will remove that mock as well.

Reply:
@AlekSi ^